docs/source/basics/tokenizers.rst

```diff
@@ -0,0 +1,239 @@
+between raw text and token IDs.
+
+.. code-block:: python
+
+    class BaseTokenizer(Protocol):
+
+        def encode(self, text: str, **kwargs: Dict[str, Any]) -> List[int]:
+            """
+            Given a string, return the encoded list of token ids.
+
+            Args:
+                text (str): The text to encode.
+                **kwargs (Dict[str, Any]): kwargs.
+
+            Returns:
+                List[int]: The encoded list of token ids.
+            """
+            pass
+
+        def decode(self, token_ids: List[int], **kwargs: Dict[str, Any]) -> str:
+            """
+            Given a list of token ids, return the decoded text, optionally including special tokens.
+
+            Args:
+                token_ids (List[int]): The list of token ids to decode.
+                **kwargs (Dict[str, Any]): kwargs.
+
+            Returns:
+                str: The decoded text.
+            """
+            pass
+
+If you load any :class:`~torchtune.modules.tokenizers.ModelTokenizer`, you can see that it calls its underlying :class:`~torchtune.modules.tokenizers.BaseTokenizer`
+to do the actual encoding and decoding.
+
+.. code-block:: python
+
+    from torchtune.models.mistral import mistral_tokenizer
+    from torchtune.modules.tokenizers import SentencePieceBaseTokenizer
+
+    m_tokenizer = mistral_tokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+    # Mistral uses SentencePiece for its underlying BPE
+    sp_tokenizer = SentencePieceBaseTokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+
+    text = "hello world"
+
+    print(m_tokenizer.encode(text))
+    # [1, 6312, 28709, 1526, 2]
+
+    print(sp_tokenizer.encode(text))
+    # [1, 6312, 28709, 1526, 2]
+
+
+Model tokenizers
+----------------
+
+:class:`~torchtune.modules.tokenizers.ModelTokenizer` are specific to a particular model. They are required to implement the ``tokenize_messages`` method,
+which converts a list of Messages into a list of token IDs.
+
+.. code-block:: python
+
+    class ModelTokenizer(Protocol):
+
+        special_tokens: Dict[str, int]
+        max_seq_len: Optional[int]
+
+        def tokenize_messages(
+            self, messages: List[Message], **kwargs: Dict[str, Any]
+        ) -> Tuple[List[int], List[bool]]:
+            """
+            Given a list of messages, return a list of tokens and list of masks for
+            the concatenated and formatted messages.
+
+            Args:
+                messages (List[Message]): The list of messages to tokenize.
+                **kwargs (Dict[str, Any]): kwargs.
+
+            Returns:
+                Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+            """
+            pass
+
+The reason they are model specific and different from :class:`~torchtune.modules.tokenizers.BaseTokenizer`
+is because they add all the necessary special tokens or prompt templates required to prompt the model.
```

Comment:
Imo this sentence is not clearly articulated enough given how important it is. I would even frame as a question: "How are model tokenizers different from base tokenizers?" in a subsection and then go through your example below in a more detailed manner

Reply:
gonna do this in a follow up